src/qilib/utils/serialization.py

```
@@ -18,61 +18,184 @@
+        self.decoder = Decoder()
+
+        if encoders is None:
+            encoders = {}
+        self.encoder.encoders = encoders
+        if decoders is None:
+            decoders = {}
+        self.decoder.decoders = decoders
+
+        self.register_encoder(bytes, _encode_bytes)
+        self.register_decoder(bytes.__name__, _decode_bytes)
+        self.register_encoder(np.ndarray, MongoDataSetIO.encode_numpy_array)
+        self.register_decoder(np.array.__name__, MongoDataSetIO.decode_numpy_array)
+
+    def register_encoder(self, type_: type, encode_func: TransformFunction):
+        """ Registers an encoder for a given type
+
+        Args:
+            type_: The type
+            encode_func: The transform function
+
+        """
+
+        self.encoder.encoders[type_] = encode_func
+
+    def register_decoder(self, type_name: str, decode_func: TransformFunction):
+        """ Registers a decoder for a given type
+
+        Args:
+            type_name: The type name
+            decode_func: The transform function
+        """
+
+        self.decoder.decoders[type_name] = decode_func
+
+    def serialize(self, data):
+        """ Serializes a Python object to JSON
+
+        Args:
+            data: Any Python object
+
+        Returns:
+            JSON encoded string
+
+        """
+
+        return self.encoder.encode(self.transform_data(data))
+
+    def unserialize(self, data):
+        """ Unserializes a JSON string to a Python object
+
+        Args:
+            data: The JSON encoded string
+
+        Returns:
+            A Python object decoded from the JSON string
+        """
+
+        return self.decoder.decode(data)
+
+    def transform_data(self, data: Any) -> Any:
+        """ Recursively transfer a Python object and apply transform functions to it
+
+        Args:
+            data: Any Python object
+
+        Returns:
+            The transformed data
+        """
+
+        if isinstance(data, dict):
```

```
import qilib
from collections import OrderedDict
a=OrderedDict({'s':1, 'a': 2})

j=qilib.utils.serialization.serialize(a)
print(j)
{"s": 1, "a": 2}
```
So an `OrderedDict` is silently converted to a `dict`. I cannot prevent this by adding a custom encoder, since that is only applied after this check. This might be ok for qilib, but I think it will not work for qtt (will check that later).